File: scooter-control/button/_button_handlers/_button_handler_base.py
# Python Imports
from abc import ABC, abstractmethod

# Main
class ButtonHandlerBase(ABC):
    def __init__(self) -> None:
        self._enabled: bool = True

    def enable(self) -> None:
        self._enabled = True
        self.on_enable()
    def disable(self) -> None:
        self._enabled = False
        self.on_disable()
    @property
    def enabled(self) -> bool:
        return self._enabled
    
    @abstractmethod
    def on_enable(self) -> None:
        pass
    @abstractmethod
    def on_disable(self) -> None:
        pass

    @abstractmethod
    def on_value_raised(self) -> None:
        pass
    @abstractmethod
    def on_value_recessed(self) -> None:
        pass

    @abstractmethod
    def update(self) -> None:
        pass

File: scooter-control/libs/scootercontrol/button/handlers/ButtonHandlerBase.cpp
// Class definition
#include "ButtonHandlerBase.hpp"

void ButtonHandlerBase::enable() {
    if (!_enabled) {
        _enabled = true;
        _onEnabled();
    }
}
void ButtonHandlerBase::disable() {
    if (_enabled) {
        _enabled = false;
        _onDisabled();
    }
}
void ButtonHandlerBase::setState(bool enabled) {
    if (enabled) {
        enable();
    } else {
        disable();
    }
}
bool ButtonHandlerBase::isEnabled() {
    return _enabled;
}
